src/ra_snapshot.erl

```diff
@@ -235,37 +233,87 @@ find_checkpoints(#?MODULE{uid = UId,
-                      case Module:validate(CP) of
-                          ok ->
-                              {ok, #{index := Idx, term := Term}} =
-                              Module:read_meta(CP),
-                              case Idx > CurrentIdx of
-                                  true ->
-                                      {true, {Idx, Term}};
-                                  false ->
-                                      ?INFO("ra_snapshot: ~ts: removing "
-                                            "checkpoint ~s as was older than the "
-                                            "current snapshot.",
-                                            [UId, CP]),
-                                      delete(CheckpointDir, {Idx, Term}),
-                                      false
-                              end;
-                          Err ->
-                              ?INFO("ra_snapshot: ~ts: removing checkpoint ~s as "
-                                    "did not validate. Err: ~w",
-                                    [UId, CP, Err]),
-                              ra_lib:recursive_delete(CP),
-                              false
-                      end
-              end, CPFiles),
-            State#?MODULE{checkpoints = Checkpoints}
+            find_checkpoints(CPFiles, State, CurrentIdx, [])
     end.
 
+find_checkpoints([], State, _CurrentIdx, Checkpoints) ->
+    %% Reverse so that the most recent checkpoints come first.
+    State#?MODULE{checkpoints = lists:reverse(Checkpoints)};
+find_checkpoints(
+  [File | Files],
+  #?MODULE{uid = UId,
+           module = Module,
+           checkpoint_directory = CheckpointDir} = State,
+  CurrentIdx, []) ->
+    %% When we haven't yet found a valid checkpoint (`Checkpoints =:= []`),
+    %% fully validate the file with the `ra_snapshot:validate/1` callback to
+    %% ensure that we can recover from the latest checkpoint.
+    CP = filename:join(CheckpointDir, File),
+    case Module:validate(CP) of
+        ok ->
+            {ok, #{index := Idx, term := Term}} = Module:read_meta(CP),
+            case Idx > CurrentIdx of
+                true ->
+                    find_checkpoints(Files, State, CurrentIdx, [{Idx, Term}]);
+                false ->
+                    %% If the first valid checkpoint is older than the snapshot
+                    %% index then all checkpoints in `Files` are older as well.
+                    %% Delete all checkpoints and bail.
+                    delete_stale_checkpoints(
+                      UId, CheckpointDir, [File | Files]),
+                    State
+            end;
+        Err ->
+            ?INFO("ra_snapshot: ~ts: removing checkpoint ~s as it did not "
+                  "validate. Err: ~w",
+                  [UId, CP, Err]),
+            _ = ra_lib:recursive_delete(CP),
+            find_checkpoints(Files, State, CurrentIdx, [])
+    end;
+find_checkpoints(
```

Comment:
same as the other clause, the first arg can go on the same line as the function.